web/Controllers/PropertyController.cs

```diff
@@ -151,7 +139,55 @@ public ActionResult ViewProperty(int propertyId)
             _logger.LogWarning("Property with ID {PropertyId} does not exist", propertyId);
             return RedirectToAction("Error", "Home", new { status = 404 });
         }
-        var propertyViewModel = PropertyViewModel.FromDbModel(model);
+        PropertyViewModel propertyViewModel = PropertyViewModel.FromDbModel(model);
         return View(propertyViewModel);
     }
+
+    [HttpGet]
+    public async Task<IActionResult> ListPropertyImages(int propertyId)
+    {
+        List<string> fileNames = await _azureStorage.ListFilesAsync("property", propertyId);
+        return View(fileNames);
+    }
+    
+    [Authorize(Roles = "Landlord, Admin")]
+    [HttpPost("addImage")]
+    public async Task<IActionResult> AddPropertyImages(List<IFormFile> images, int propertyId)
+    {
+        var user = GetCurrentUser();
+        if (!user.IsAdmin && _propertyService.GetPropertyByPropertyId(propertyId)!.LandlordId != user.LandlordId)
+        {
+            return StatusCode(403);
+        }
+        foreach (var image in images)
+        {
+            if (image.Length > 0)
+            {
+                await _azureStorage.UploadFileAsync(image, "property", propertyId);
+            }
+        }
+        return RedirectToAction("ListPropertyImages", "Property", new{propertyId});
+    }
+    
+    [HttpPost("displayImage")]
+    public async Task<IActionResult> DisplayPropertyImage(int propertyId, string fileName)
+    {
+        var image = await _azureStorage.DownloadFileAsync("property", propertyId, fileName);

```

Comment:
Should check permissions here too (probably want a helper method!)

Reply:
Surely the point of uploading images is so that they can be seen by potential tenants, and therefore we shouldn't check permissions here? Hence why I also didn't put a permissions check on ListAllProperties, so that these images would be accessible via that list for potential tenants to view.
There should probably be some permissions settings on the View so that the Add and Delete buttons only appear to the correct users, but I anticipate that 'Display' will end up on some publicly available page while 'Add' and 'Delete' will end up on an Edit page (which has its own permissions checks), hence effectively negating the need for that logic within the view itself.

Reply:
For now we should be checking permissions and info about a property should only be available to its landlord and admins.

In the long run yes, and I hope we'll get to start implementing that tenants viewing pages - however I think it will still be fairly restricted, with tenants just being able to see properties they've been matched with and linked to rather than browsing the site like RightMove.